File: parsers/java/src/main/java/com/aperturesyndicate/synx/SynxCalc.java
package com.aperturesyndicate.synx;

import java.util.ArrayList;
import java.util.List;

/** Safe arithmetic evaluator for {@code :calc}. Mirrors {@code crates/synx-core/src/calc.rs}. */
public final class SynxCalc {

    private SynxCalc() {}

    public static final class Result {
        public final boolean ok;
        public final double value;
        public final String error;
        Result(boolean ok, double value, String error) {
            this.ok = ok; this.value = value; this.error = error;
        }
        public static Result success(double v) { return new Result(true, v, ""); }
        public static Result failure(String msg) { return new Result(false, 0, msg); }
    }

    public static Result evaluate(String expr) {
        String t = expr == null ? "" : expr.strip();
        if (t.isEmpty()) return Result.success(0);
        List<Token> tokens = new ArrayList<>();
        String err = tokenize(t, tokens);
        if (err != null) return Result.failure(err);
        if (tokens.isEmpty()) return Result.success(0);
        return new ExprParser(tokens).parse();
    }

    // ─── tokens ─────────────────────────────────────────────────────────────

    private enum Kind { NUMBER, OP, LPAREN, RPAREN }

    private static final class Token {
        final Kind kind;
        final double number;
        final char op;
        Token(Kind k, double n, char op) { this.kind = k; this.number = n; this.op = op; }
    }

    private static String tokenize(String expr, List<Token> tokens) {
        int i = 0;
        int len = expr.length();
        while (i < len) {
            char c = expr.charAt(i);
            if (c == ' ' || c == '\t') { i++; continue; }
            boolean isDigit = c >= '0' && c <= '9';
            boolean isDotNum = c == '.' && i + 1 < len && expr.charAt(i + 1) >= '0' && expr.charAt(i + 1) <= '9';
            boolean isUnary = false;
            if (c == '-') {
                if (tokens.isEmpty()) isUnary = true;
                else {
                    Token last = tokens.get(tokens.size() - 1);
                    if (last.kind == Kind.OP || last.kind == Kind.LPAREN) isUnary = true;
                }
            }
            if (isDigit || isDotNum || isUnary) {
                int start = i;
                if (c == '-') i++;
                while (i < len) {
                    char x = expr.charAt(i);
                    if ((x >= '0' && x <= '9') || x == '.') i++;
                    else break;
                }
                String s = expr.substring(start, i);
                try {
                    tokens.add(new Token(Kind.NUMBER, Double.parseDouble(s), '\0'));
                } catch (NumberFormatException e) {
                    return "SYNX :calc - invalid number: '" + s + "'";
                }
                continue;
            }
            if (c == '+' || c == '-' || c == '*' || c == '/' || c == '%') {
                tokens.add(new Token(Kind.OP, 0, c)); i++; continue;
            }
            if (c == '(') { tokens.add(new Token(Kind.LPAREN, 0, '\0')); i++; continue; }
            if (c == ')') { tokens.add(new Token(Kind.RPAREN, 0, '\0')); i++; continue; }
            return "SYNX :calc - unexpected character: '" + c + "'";
        }
        return null;
    }

    private static final class ExprParser {
        final List<Token> tokens;
        int pos;
        ExprParser(List<Token> t) { this.tokens = t; }

        Result parse() {
            double[] out = { 0 };
            String e = expr(out);
            if (e != null) return Result.failure(e);
            if (pos < tokens.size()) {
                return Result.failure("SYNX :calc - unexpected token at position " + pos);
            }
            return Result.success(out[0]);
        }

        String expr(double[] out) {
            double[] left = { 0 };
            String e = term(left);
            if (e != null) return e;
            while (pos < tokens.size()) {
                Token t = tokens.get(pos);
                if (t.kind == Kind.OP && t.op == '+') {
                    pos++;
                    double[] r = { 0 };
                    e = term(r); if (e != null) return e;
                    left[0] += r[0];
                } else if (t.kind == Kind.OP && t.op == '-') {
                    pos++;
                    double[] r = { 0 };
                    e = term(r); if (e != null) return e;
                    left[0] -= r[0];
                } else break;
            }
            out[0] = left[0];
            return null;
        }

        String term(double[] out) {
            double[] left = { 0 };
            String e = factor(left);
            if (e != null) return e;
            while (pos < tokens.size()) {
                Token t = tokens.get(pos);
                if (t.kind != Kind.OP) break;
                switch (t.op) {
                    case '*': {
                        pos++;
                        double[] r = { 0 };
                        e = factor(r); if (e != null) return e;
                        left[0] *= r[0]; break;
                    }
                    case '/': {
                        pos++;
                        double[] r = { 0 };
                        e = factor(r); if (e != null) return e;
                        if (r[0] == 0) return "SYNX :calc - division by zero";
                        left[0] /= r[0]; break;
                    }
                    case '%': {
                        pos++;
                        double[] r = { 0 };
                        e = factor(r); if (e != null) return e;
                        if (r[0] == 0) return "SYNX :calc - division by zero";
                        left[0] = left[0] % r[0]; break;
                    }
                    default: out[0] = left[0]; return null;
                }
            }
            out[0] = left[0];
            return null;
        }

        String factor(double[] out) {
            if (pos >= tokens.size()) return "SYNX :calc - unexpected end of expression";
            Token t = tokens.get(pos);
            if (t.kind == Kind.NUMBER) { pos++; out[0] = t.number; return null; }
            if (t.kind == Kind.LPAREN) {
                pos++;
                double[] v = { 0 };
                String e = expr(v);
                if (e != null) return e;
                if (pos >= tokens.size()) return "SYNX :calc - missing closing parenthesis";
                if (tokens.get(pos).kind != Kind.RPAREN) {
                    return "SYNX :calc - missing closing parenthesis";
                }
                pos++; out[0] = v[0]; return null;
            }
            return "SYNX :calc - unexpected token";
        }
    }
}

File: parsers/swift/Sources/Synx/SynxCalc.swift
// Safe arithmetic evaluator for `:calc`. Mirrors crates/synx-core/src/calc.rs.
import Foundation

public enum SynxCalc {

    public struct Result {
        public let ok: Bool
        public let value: Double
        public let error: String
    }

    public static func evaluate(_ expr: String) -> Result {
        let trimmed = expr.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty { return Result(ok: true, value: 0, error: "") }
        var tokens: [Token] = []
        if let err = tokenize(trimmed, into: &tokens) {
            return Result(ok: false, value: 0, error: err)
        }
        if tokens.isEmpty { return Result(ok: true, value: 0, error: "") }
        var parser = ExprParser(tokens: tokens)
        return parser.parse()
    }

    // MARK: - tokens

    private enum Token {
        case number(Double)
        case op(Character)
        case lparen
        case rparen
    }

    private static func tokenize(_ expr: String, into tokens: inout [Token]) -> String? {
        let chars = Array(expr)
        var i = 0
        while i < chars.count {
            let c = chars[i]
            if c == " " || c == "\t" { i += 1; continue }

            let isDigit = (c >= "0" && c <= "9")
            let isDotNumber = (c == "." && i + 1 < chars.count
                               && chars[i + 1] >= "0" && chars[i + 1] <= "9")
            let isUnaryMinus: Bool = {
                if c != "-" { return false }
                if tokens.isEmpty { return true }
                switch tokens.last! {
                case .op, .lparen: return true
                default: return false
                }
            }()
            if isDigit || isDotNumber || isUnaryMinus {
                let start = i
                if c == "-" { i += 1 }
                while i < chars.count && (chars[i] >= "0" && chars[i] <= "9" || chars[i] == ".") {
                    i += 1
                }
                let str = String(chars[start..<i])
                guard let d = Double(str) else {
                    return "SYNX :calc - invalid number: '\(str)'"
                }
                tokens.append(.number(d))
                continue
            }
            if c == "+" || c == "-" || c == "*" || c == "/" || c == "%" {
                tokens.append(.op(c)); i += 1; continue
            }
            if c == "(" { tokens.append(.lparen); i += 1; continue }
            if c == ")" { tokens.append(.rparen); i += 1; continue }
            return "SYNX :calc - unexpected character: '\(c)'"
        }
        return nil
    }

    private struct ExprParser {
        let tokens: [Token]
        var pos: Int = 0

        mutating func parse() -> Result {
            var v: Double = 0
            if let err = expr(&v) { return Result(ok: false, value: 0, error: err) }
            if pos < tokens.count {
                return Result(ok: false, value: 0,
                              error: "SYNX :calc - unexpected token at position \(pos)")
            }
            return Result(ok: true, value: v, error: "")
        }

        mutating func expr(_ out: inout Double) -> String? {
            var left: Double = 0
            if let e = term(&left) { return e }
            while pos < tokens.count {
                if case .op(let c) = tokens[pos], c == "+" {
                    pos += 1
                    var r: Double = 0
                    if let e = term(&r) { return e }
                    left += r
                } else if case .op(let c) = tokens[pos], c == "-" {
                    pos += 1
                    var r: Double = 0
                    if let e = term(&r) { return e }
                    left -= r
                } else { break }
            }
            out = left
            return nil
        }

        mutating func term(_ out: inout Double) -> String? {
            var left: Double = 0
            if let e = factor(&left) { return e }
            while pos < tokens.count {
                if case .op(let c) = tokens[pos] {
                    switch c {
                    case "*":
                        pos += 1
                        var r: Double = 0
                        if let e = factor(&r) { return e }
                        left *= r
                    case "/":
                        pos += 1
                        var r: Double = 0
                        if let e = factor(&r) { return e }
                        if r == 0 { return "SYNX :calc - division by zero" }
                        left /= r
                    case "%":
                        pos += 1
                        var r: Double = 0
                        if let e = factor(&r) { return e }
                        if r == 0 { return "SYNX :calc - division by zero" }
                        left = left.truncatingRemainder(dividingBy: r)
                    default: out = left; return nil
                    }
                } else { break }
            }
            out = left
            return nil
        }

        mutating func factor(_ out: inout Double) -> String? {
            if pos >= tokens.count { return "SYNX :calc - unexpected end of expression" }
            switch tokens[pos] {
            case .number(let n):
                pos += 1
                out = n
                return nil
            case .lparen:
                pos += 1
                var v: Double = 0
                if let e = expr(&v) { return e }
                if pos >= tokens.count {
                    return "SYNX :calc - missing closing parenthesis"
                }
                if case .rparen = tokens[pos] { pos += 1; out = v; return nil }
                return "SYNX :calc - missing closing parenthesis"
            default:
                return "SYNX :calc - unexpected token"
            }
        }
    }
}
